### PR TITLE
create the "secretkey" command

### PR DESCRIPTION
**Pull request by oconnor663**
This command takes either "encryption" or "signing" as an argument and
prints the corresponding device secret key to the terminal, as hex. It's
not for public use, and we keep the usage string undefined so that it
doesn't show up in help.

I need this to test the Python implementation of SaltPack, by encrypting messages on the Go side and decrypting them on the Python side. I use this command to get the encryption secret keys for my test users.

@patrickxb @maxtaco Let me know if you think there's any security concern here. This makes it possible for any user process to request the secret keys from the service. One option if we don't like it would be to make this dev-mode-only, or to require you to start the daemon with a specific flag. (But of course, an adversary who can run code on your machine can work around all of these things.)

### CODE REVIEW

**Comment by patrickxb** [comment]
One thing you could do if you were concerned about it is to clear the stream cache 

     e.G().LoginState().Account(func (a *Account) {
         a.ClearStreamCache()
     }, "clear stream cache")

and then use GetSecretKeyWithPassphrase.

That would force the user of the command to enter the passphrase to unlock the secret keys.

File: go/client/cmd_secretkey.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+func (c *CmdSecretKey) Run() (err error) {
+	cli, err := GetSecretKeysClient(c.G())
+	if err != nil {
+		return err
+	}
+
+	protocols := []rpc.Protocol{
+		NewSecretUIProtocol(G),
+	}
+	if err = RegisterProtocols(protocols); err != nil {
```
Comment (maxtaco) [inline go/client/cmd_secretkey.go]: Please use the contextified version!

File: go/client/cmd_secretkey.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func (c *CmdSecretKey) Run() (err error) {
+	cli, err := GetSecretKeysClient(c.G())
+	if err != nil {
+		return err
+	}
+
+	protocols := []rpc.Protocol{
+		NewSecretUIProtocol(G),
```
Comment (maxtaco) [inline go/client/cmd_secretkey.go]: Please use a local `G`

File: go/service/secretkeys.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+func NewSecretKeysHandler(xp rpc.Transporter, g *libkb.GlobalContext) *SecretKeysHandler {
+	return &SecretKeysHandler{
+		BaseHandler:  NewBaseHandler(xp),
+		Contextified: libkb.NewContextified(g),
+	}
+}
+
+// BTC creates a BTCEngine and runs it.
```
Comment (maxtaco) [inline go/service/secretkeys.go]: bad comment here

File: protocol/avdl/secretkeys.avdl [hunk 4]
```diff
@@ -0,0 +1,12 @@
+@namespace("keybase.1")
+
+protocol SecretKeys {
+  import idl "common.avdl";
+
+  record SecretKeys {
+    string signingSecretKey;
```
Comment (maxtaco) [inline protocol/avdl/secretkeys.avdl]: is there a better type than string here?

File: go/client/cmd_secretkey.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	"golang.org/x/net/context"
+)
+
+type CmdSecretKey struct {
+	libkb.Contextified
+	keytype string
+}
+
+func (c *CmdSecretKey) ParseArgv(ctx *cli.Context) error {
+	if len(ctx.Args()) != 1 {
```
Comment (maxtaco) [inline go/client/cmd_secretkey.go]: Isn't this better to do with a flag?

**Comment by maxtaco** [comment]
made a few comments

File: go/service/secretkeys.go (hunk 3 above)
Comment (oconnor663) [inline go/service/secretkeys.go]: Ah shoot, lots of copypasta errors on my part.

File: go/client/cmd_secretkey.go (hunk 5 above)
Comment (oconnor663) [inline go/client/cmd_secretkey.go]: I prefer not to use flags for arguments that are mandatory. When I see

    keybase secretkey --signing

I tend to assume that `keybase secretkey` is valid by itself, which it isn't. An alternative might be to define two commands, maybe `signing-key` and `encryption-key` or something. Thoughts?

File: protocol/avdl/secretkeys.avdl (hunk 4 above)
Comment (oconnor663) [inline protocol/avdl/secretkeys.avdl]: I'll define a couple new types that are just bytes.

File: go/client/cmd_secretkey.go (hunk 5 above)
Comment (maxtaco) [inline go/client/cmd_secretkey.go]: in this case, let's stet

File: protocol/avdl/secretkeys.avdl (hunk 4 above)
Comment (maxtaco) [inline protocol/avdl/secretkeys.avdl]: ok, there might already be something here.  just common.avdl

Comment (oconnor663) [inline protocol/avdl/secretkeys.avdl]: I think we used to, before we turned KIDs back into strings. Easy to add.

**Comment by oconnor663** [comment]
Addressed Max's comments. @patrickxb, when I add the stanza you suggested, I still don't get prompted for my password after the first time. Is there another piece of cache we would need to clear?

File: protocol/avdl/secretkeys.avdl [hunk 6]
```diff
@@ -0,0 +1,17 @@
+@namespace("keybase.1")
+
+protocol SecretKeys {
+  import idl "common.avdl";
+
+  fixed NaclSigningKeyPublic(32);
+  fixed NaclSigningKeyPrivate(64);
+  fixed NaclDHKeyPublic(32);
+  fixed NaclDHKeyPrivate(32);
```
Comment (maxtaco) [inline protocol/avdl/secretkeys.avdl]: i guess it's good we weren't shipping these over RPC previously...

**Comment by patrickxb** [comment]
@oconnor663 also call a.ClearCachedSecretKeys() after a.ClearStreamCache().

**Comment by oconnor663** [comment]
@patrickxb figured out that we needed to define an additional function, `Account.ClearKeyring()`.

**Comment by patrickxb** [comment]
LGTM...will be nice to have this command!

**Pull request merged**